safehttp/task.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+)
+
+type Task struct {
+	rw  http.ResponseWriter
+	req *IncomingRequest
+
+	cfg HandlerConfig
+
+	code   StatusCode
+	header Header
+
+	written      bool
```

Comment:
We are not distinguishing between actual writes and write calls. This means that when we start writing an error we don't know if the response serialization was already started (and thus writing the error is useless).

Reply:
Yup, I'm gonna probably add a `dispatched` boolean here. I'll leave a TODO to not cram it into this PR.